### PR TITLE
fix: L02, M01, M04, M05, M07, M08, C01, C02

### PR DESCRIPTION
**Pull request by fedealconada**
Fixes L02, M01, M04, M05, M07, M08, C01, C02,

### CODE REVIEW

File: src/bridger/Bridger.sol [hunk 1]
```diff
@@ -64,6 +64,7 @@ contract Bridger is
     bytes32 public immutable override domainSeparator;
     address public immutable override l2Vault;
     address public override senderAccount;
+    address public override exchangeProxy;
```
Comment (rrecuero) [inline src/bridger/Bridger.sol:67]: immutable

File: src/bridger/Bridger.sol [hunk 2]
```diff
@@ -101,6 +102,7 @@ contract Bridger is
         _transferOwnership(msg.sender);
         swapsEnabled = false;
         senderAccount = _senderAccount;
+        exchangeProxy = 0xDef1C0ded9bec7F1a1670819833240f027b25EfF;
```
Comment (rrecuero) [inline src/bridger/Bridger.sol:105]: set in constructor when immutable

File: src/bridger/Bridger.sol (hunk 1 above)
Comment (rrecuero) [inline src/bridger/Bridger.sol:67]: immutable to save gas

File: src/bridger/Bridger.sol [hunk 3]
```diff
@@ -149,15 +151,16 @@ contract Bridger is
...
         IBridger.SwapData calldata _swapData,
         bytes calldata _permitSignature
-    ) external payable override whenNotPaused onlySignerVerified(_signatureData) onlyPrivileged {
+    ) external payable override whenNotPaused nonReentrant onlyPrivileged onlySignerVerified(_signatureData) {
         _isFinalAssetAllowed(_signatureData.finalAsset);
-        // Only allows assets for now which do not require swap.
         if (_signatureData.inputAsset != _signatureData.finalAsset && !allowedAssets[_signatureData.inputAsset]) {
             // checks for USDe special case
-            if (_signatureData.inputAsset != USDe && _signatureData.finalAsset != sUSDe) {
+            if (_signatureData.inputAsset != USDe || _signatureData.finalAsset != sUSDe) {
```
Comment (rrecuero) [inline src/bridger/Bridger.sol:158]: why did you change this? I believe the condition was right before

File: src/bridger/Bridger.sol [hunk 4]
```diff
@@ -308,9 +310,9 @@ contract Bridger is
         amountBought = ERC20(wstETH).balanceOf(address(this)) - balanceBefore;
     }
 
-    function _stakeAssets(address _asset, uint256 _amount) private {
+    function _stakeAssets(address _asset, uint256 _amount) private returns (uint256) {
         IERC20(_asset).approve(address(sUSDe), _amount);
-        IsUSDe(sUSDe).deposit(_amount, address(this));
+        return IsUSDe(sUSDe).deposit(_amount, address(this));
```
Comment (rrecuero) [inline src/bridger/Bridger.sol:315]: I dont think this will work, because it will accumulate the amount on each deposit.

I recommend just calling balanceOf before and after and using the delta

Comment (rrecuero) [reply src/bridger/Bridger.sol:315]: or maybe better to multiply the amount staked for the exchange rate, the contract may have it

Comment (fedealconada) [reply src/bridger/Bridger.sol:315]: >  because it will accumulate the amount on each deposit

AFAIK, it does not accumulate, it just returns the value of shared that you got by depositing a given amount

> or maybe better to multiply the amount staked for the exchange rate, the contract may have it

`deposit` calls `previewDeposit` which returns the amount of shares by calling `_convertToShares` which kind of does the multiplication you are suggesting

**Comment by fedealconada** [comment]
added fixes for M5 and M8 cc @rrecuero

File: src/bridger/Bridger.sol [hunk 5]
```diff
@@ -140,25 +144,27 @@ contract Bridger is
...
+        nonReentrant
+        onlyPrivileged
+        onlySignerVerified(_signatureData)
+    {
         _isFinalAssetAllowed(_signatureData.finalAsset);
-        // Only allows assets for now which do not require swap.
         if (_signatureData.inputAsset != _signatureData.finalAsset && !allowedAssets[_signatureData.inputAsset]) {
             // checks for USDe special case
-            if (_signatureData.inputAsset != USDe && _signatureData.finalAsset != sUSDe) {
+            if (_signatureData.inputAsset != USDe || _signatureData.finalAsset != sUSDe) {
```
Comment (ylv-io) [inline src/bridger/Bridger.sol:162]:
```suggestion
            if (_signatureData.inputAsset == USDe && _signatureData.finalAsset != sUSDe) {
```
Should it be this? Probably need a test case for this.

File: .github/workflows/forge-test.yml [hunk 6]
```diff
@@ -52,7 +52,7 @@ jobs:
       - uses: ./.github/actions/install
 
       - name: Run tests
-        run: forge test --gas-report > gasreport.ansi
```
Comment (fedealconada) [inline .github/workflows/forge-test.yml:55]: temporarily removing this until foundry fixes some issue related to this

File: src/bridger/Bridger.sol (hunk 5 above)
Comment (rrecuero) [reply src/bridger/Bridger.sol:162]: It has liquidity in matcha so we can allow other final assets

File: src/bridger/Bridger.sol (hunk 4 above)
Comment (rrecuero) [reply src/bridger/Bridger.sol:320]: perfect, sounds good

File: src/interfaces/IBridger.sol [hunk 7]
```diff
@@ -37,20 +37,22 @@ interface IBridger {
...
+        address kintoWallet; // Kinto Wallet Address on L2 where tokens will be deposited
         address signer;
         address inputAsset;
-        uint256 amount;
         address finalAsset;
+        uint256 amount;
         uint256 nonce;
         uint256 expiresAt;
         bytes signature;
+        SwapData swapData; // Struct with all the required information to swap the tokens
```
Comment (rrecuero) [inline src/interfaces/IBridger.sol:55]: It can take a while since the user signs and the signature is sent. swapData cannot be here but I would include minAmount from swapDta in here

File: src/interfaces/IBridger.sol [hunk 8]
```diff
@@ -65,13 +65,12 @@ interface IBridger {
...
+        SwapParams swapParams;
     }
 
-    struct SwapData {
+    struct SwapParams {
         address spender;
         address swapTarget;
-        bytes swapCallData;
         uint256 gasFee;
         uint256 minReceive;
```
Comment (rrecuero) [inline src/interfaces/IBridger.sol:75]: this needs to be moved to the signature

File: src/interfaces/IBridger.sol [hunk 9]
```diff
@@ -65,13 +65,12 @@ interface IBridger {
         uint256 nonce;
         uint256 expiresAt;
         bytes signature;
-        SwapData swapData; // Struct with all the required information to swap the tokens
+        SwapParams swapParams;
     }
 
-    struct SwapData {
+    struct SwapParams {
         address spender;
```
Comment (rrecuero) [inline src/interfaces/IBridger.sol:72]: swapCallData can stay here

File: src/interfaces/IBridger.sol [hunk 10]
```diff
@@ -65,13 +65,12 @@ interface IBridger {
         uint256 nonce;
         uint256 expiresAt;
         bytes signature;
-        SwapData swapData; // Struct with all the required information to swap the tokens
+        SwapParams swapParams;
```
Comment (rrecuero) [inline src/interfaces/IBridger.sol:68]: move this out of here but put minReceiveAmount here

**Review by rrecuero — APPROVED** [review]
